### PR TITLE
Add `repair-brig-clients-table` to clean up after the fix in #3504.

### PR DESCRIPTION
**Pull request by fisx**
https://wearezeta.atlassian.net/browse/WPB-3888

```
make c package=repair-brig-clients-table
./dist/repair-brig-clients-table --dry-run --cassandra-port-brig 9044 --cassandra-keyspace-brig brig > repair-brig-clients-table.log 2>&1
```

Without `--dry-run` this will actually delete the offending entries.

I will attach the log as soon as it's done (currently at 7M clients and 30 of them having all fields NULL except keys and last_active).

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

**Comment by fisx** [comment]
```
$ cat repair-brig-clients-table.log | grep '\*\*\*' | wc
     36     360    6323

$ cat repair-brig-clients-table.log | wc
   9013   54166  375779
```

[repair-brig-clients-table.log](https://github.com/wireapp/wire-server/files/12361258/repair-brig-clients-table.log)

so, 36 broken clients.  i suggest we remove them?

**Comment by fisx** [comment]
I think this is ready to merge (whether integration tests have passed or not...)

File: tools/db/repair-brig-clients-table/repair-brig-clients-table.cabal [hunk 1]
```diff
@@ -0,0 +1,83 @@
+cabal-version: 1.12
+name:          repair-brig-clients-table
+version:       1.0.0
+synopsis:
+  Removes and reports entries from brig.clients that have been accidentally upserted.
+
+category:      Network
+author:        Wire Swiss GmbH
+maintainer:    Wire Swiss GmbH <backend@wire.com>
+copyright:     (c) 2018 Wire Swiss GmbH
```
Comment (supersven) [inline tools/db/repair-brig-clients-table/repair-brig-clients-table.cabal:10]:
```suggestion
copyright:     (c) 2023 Wire Swiss GmbH
```

File: tools/db/repair-brig-clients-table/repair-brig-clients-table.cabal [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+executable repair-brig-clients-table
+  main-is:            Main.hs
+  other-modules:
+    Options
+    Paths_repair_brig_clients_table
+    Work
+
+  hs-source-dirs:     src
+  default-extensions:
```
Comment (supersven) [inline tools/db/repair-brig-clients-table/repair-brig-clients-table.cabal:22]: I think there's some overlapping with what `GHC2021` brings... However, this is just a one-shot cleanup tool. :man_shrugging:

File: tools/db/repair-brig-clients-table/src/Options.hs [hunk 3]
```diff
@@ -0,0 +1,94 @@
+{-# LANGUAGE OverloadedStrings #-}
```
Comment (supersven) [inline tools/db/repair-brig-clients-table/src/Options.hs:1]: Isn't this a default plugin, as defined in the cabal file? :thinking:

File: tools/db/repair-brig-clients-table/src/Work.hs [hunk 4]
```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE OverloadedStrings #-}
```
Comment (supersven) [inline tools/db/repair-brig-clients-table/src/Work.hs:1]: Default extension as defined in the cabal file, isn't it?

File: tools/db/repair-brig-clients-table/src/Work.hs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+{-# LANGUAGE ScopedTypeVariables #-}
```
Comment (supersven) [inline tools/db/repair-brig-clients-table/src/Work.hs:2]: Default extension as defined in the cabal file, isn't it?

File: tools/db/repair-brig-clients-table/src/Work.hs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+{-# OPTIONS_GHC -Wno-orphans -Wno-unused-imports #-}
```
Comment (supersven) [inline tools/db/repair-brig-clients-table/src/Work.hs:3]: Are these options needed here at all? :thinking:

File: tools/db/repair-brig-clients-table/src/Work.hs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import Data.Time.Clock
+import Imports
+import System.Logger (Logger)
+import System.Logger qualified as Log
+import UnliftIO.Async (pooledMapConcurrentlyN)
+import Wire.API.Team.Feature
+import Wire.API.User
+
+runCommand :: Bool -> Logger -> ClientState -> IO ()
+runCommand dryRun l brig = do
```
Comment (supersven) [inline tools/db/repair-brig-clients-table/src/Work.hs:39]:
```suggestion
runCommand dryRun l brig =
```

Isn't the `do` superfluous? :thinking:

File: tools/db/repair-brig-clients-table/src/Work.hs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  if dryRun
+    then do
+      Log.info l (Log.msg $ "would run: " <> rmqs)
+    else do
+      Log.info l (Log.msg $ "running: " <> rmqs)
+      rm user client
+      Log.info l (Log.msg @Text "removed!")
+  where
+    rm :: MonadClient m => UserId -> Text -> m ()
+    rm uid cid = do
```
Comment (supersven) [inline tools/db/repair-brig-clients-table/src/Work.hs:85]:
```suggestion
    rm uid cid = 
```

Superfluous `do`? :thinking:

File: tools/db/repair-brig-clients-table/src/Work.hs [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    Maybe Text, -- model
+    Maybe UTCTime, -- tstamp
+    Maybe Int32 -- type
+  )
+
+getClients :: ConduitM () [ClientRow] Client ()
+getClients = paginateC cql (paramsP LocalQuorum () pageSize) x5
+  where
+    cql :: PrepQuery R () ClientRow
+    cql = "select user, client, capabilities, class, cookie, label, last_active, model, tstamp, type from clients"
```
Comment (supersven) [inline tools/db/repair-brig-clients-table/src/Work.hs:71]: I'm wondering if *select all clients, then filter in Haskell* is the best strategy. Can't we limit the workload by a smarter query? I.e. one that queries for `Rows` with erroneous values?

File: tools/db/repair-brig-clients-table/repair-brig-clients-table.cabal (hunk 2 above)
Comment (fisx) [reply tools/db/repair-brig-clients-table/repair-brig-clients-table.cabal:22]: yeah, i wouldn't want to spend time on this now...

File: tools/db/repair-brig-clients-table/src/Work.hs (hunk 9 above)
Comment (supersven) [reply tools/db/repair-brig-clients-table/src/Work.hs:65]: Okay, I'm used to using databases ... not these *"things"* :wink: 

As it's a one-time action, I think we don't have to spend more time on this :smile_cat:

**Comment by fisx** [comment]
CI failure is unrelated.

**Pull request merged**